### PR TITLE
Correctly treat pan/zoom events of overlapping axes.

### PR DESCRIPTION
**Pull request by raphaelquast**
Followup for #22347

### CODE REVIEW

**Comment by scottshambaugh** [comment]
It may not make sense to roll into this PR, but while you’re looking at this could you check the behavior of 3D axes as well? They have different pan/rotate/zoom handling.

**Comment by raphaelquast** [comment]
@scottshambaugh 

Sorry to disappoint you, but this PR is already going on for a very long time and I really like to get this merged in a reasonble time. Extending it with additional functionalities is unfortunately out of scope for me at the moment (also my experience with mpl3d is limited).

**Comment by ksunden** [comment]
The newly added methods/`__init__` arguments should be added to the `pyi` stub file (with type hints, which I think are mostly just `bool | Literal["auto"]` here)

**Comment by raphaelquast** [comment]
@ksunden Stub-files are updated now!

**Comment by ksunden** [comment]
I think this largely works as desired/documented. My concern is that I think it is confusing that `ax.patch.set_facecolor("none")` behaves differently to `ax.patch.set_visible(False)` here... visually they are identical, but they differ in interactivity because the `"auto"` treatment checks `patch.is_visible()`. 

I'm not sure that that should be blocking, necessarily, but I do think that transparent axes are one of the most sensible cases for overlapping axes (along with things like insets, etc) and I'm not certain that people will have used `set_visible` rather than `set_facecolor("none")`

And of course if we _do_ decide to key off of the alpha value, we would need to decide what cutoff to use:

- alpha=0 means pass events, any nonzero value absorbs
- alpha=1 absorbs events, any non-one value passes
- something in between? 0.5 as a cutoff?

I would lean towards one of the first two, I think though not sure which one. (Under the assumption we want to key off of alpha in the first palace, which I'm not totally convinced of)

**Comment by raphaelquast** [comment]
@ksunden Thanks for taking a close look on the PR!

Let me add my 2 cents to your concerns:

In my opinion, the current implementation still seems the most intuitive... to clarify:

- Axes only forward events by default if the background is **explicitly hidden** (e.g. via `ax.patch.set_visible(False)`) 
  - I think this is quite intuitive because even if the patch is set to transparent, it is _"still present in the figure"_ (e.g. changing the color would make it immediately visible) while a hidden patch only becomes visible if it is explicitly shown (e.g. via `ax.patch.set_visible(True)`) independent of all other patch-properties.
  - In short, this means that the presence of a patch directly triggers capturing of events.

Considering the effects of transparency as an additional trigger for capturing behavior, I think it would add more confusion than it solves.

- using an alpha-value other than 0 or 1 would be quite arbitrary
- in my opinion, changing the color of a patch should not interfere with any type of low-level behavior of an axes. 

In 99% of the case, people will want axes that capture events. So even though less people use `set_visible`, the ones who want to temper with the capturing behavior will find it in the docs and all other users that only use `set_facecolor("none")` (or `set_alpha(0)`) do not interfere with the capturing behavior at all.

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  (irrespective of the patch visibility)
+
+
+``ax.set_forward_navigation_events(val)`` can be used to override the
+default behaviour:
+
+- ``True``:  Forward navigation events to axes below.
+- ``False``: Execute navigation events only on this axes.
+- ``"auto"``: Use the default behaviour
+  (``True`` for axes with an invisible patch and ``False`` otherwise).
```
Comment (tacaswell) [inline galleries/examples/showcase/pan_zoom_overlap.py:23]: be explicit about what "invisible" means here (`obj.set_visible(False)`)

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -2952,6 +2952,63 @@ def _zoom_pan_handler(self, event):
...
+            f = ax.get_forward_navigation_events()
+            if f == "auto":  # (capture = patch visibility)
+                f = not ax.patch.get_visible()
+            return not f
+
+        # get all relevant axes for the event
+        axes = list(filter(_ax_filter, self.canvas.figure.get_axes()))
+
+        if len(axes) == 0:
+            return
```
Comment (tacaswell) [inline lib/matplotlib/backend_bases.py:2974]:
```suggestion
            return []
```

Even though where this is called is gated an `if axes`, we should try to maintain type stability on the return value of this method.  Someone in 5 years will thank us.

File: lib/matplotlib/backend_bases.py [hunk 3]
```diff
@@ -2952,6 +2952,63 @@ def _zoom_pan_handler(self, event):
...
+
+        if self._nav_stack() is None:
+            self.push_current()   # Set the home button to this view.
+
+        # group axes by zorder (reverse to trigger later axes first)
+        grps = dict()
```
Comment (tacaswell) [inline lib/matplotlib/backend_bases.py:2980]: use a `defaultdict`?

File: lib/matplotlib/backend_bases.py [hunk 4]
```diff
@@ -2952,6 +2952,63 @@ def _zoom_pan_handler(self, event):
             elif event.name == "button_release_event":
                 self.release_zoom(event)
 
+    def _start_event_axes_pan(self, event, method="zoom"):
```
Comment (tacaswell) [inline lib/matplotlib/backend_bases.py:2955]:
```suggestion
    def _start_event_axes_interaction(self, event, method="zoom"):
```

I had a moment of "wait, this PR does not do zoom!"

File: lib/matplotlib/backend_bases.py [hunk 5]
```diff
@@ -2952,6 +2952,63 @@ def _zoom_pan_handler(self, event):
...
+                continue
+
+            # If the inner loop was terminated with an explicit break,
+            # terminate the outer loop as well.
+            break
+
+        # avoid duplicated triggers (but keep order of list)
+        pan_axes = list(dict.fromkeys(pan_axes))
+
+        for ax in pan_axes:
```
Comment (tacaswell) [inline lib/matplotlib/backend_bases.py:3007]: Should this be moved back to the next pan method up?  I do not think we want to start pan on zoom.

File: lib/matplotlib/backend_bases.py (hunk 3 above)
Comment (raphaelquast) [reply lib/matplotlib/backend_bases.py:2980]: The very first implementation of this PR used a `defaultdict` but this was discouraged by @anntzer here: https://github.com/matplotlib/matplotlib/pull/22347#discussion_r795046200

> (one problem with defaultdict is that any __getitem__ could be inserting a new key into the dict, and it's hard to follow which ones do so; using setdefault makes this explicit.)

so I switched to using `dict.setdefault` to address the concerns. 
I'd rather not switch back again except if there's an actual reason why this would be beneficial.

File: lib/matplotlib/backend_bases.py (hunk 4 above)
Comment (raphaelquast) [reply lib/matplotlib/backend_bases.py]: True... that's still an artefact from the initial implementation which also had an explicit `_start_event_zoom`... 
Now `_start_event_axes_interaction` is used everywhere!

File: lib/matplotlib/backend_bases.py (hunk 5 above)
Comment (raphaelquast) [reply lib/matplotlib/backend_bases.py]: Also true... moved the call to `ax.start_pan` up to `ax.press_pan`!

File: galleries/examples/showcase/pan_zoom_overlap.py (hunk 1 above)
Comment (raphaelquast) [reply galleries/examples/showcase/pan_zoom_overlap.py]: added a note on what is considered a "visible patch"

**Comment by raphaelquast** [comment]
@tacaswell thanks for having a look at the PR!
I addressed all comments so this should be ready for a re-review!

File: lib/matplotlib/backend_bases.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/backend_bases.py:2979]: Fair enough.  I'm not sure I'm fully convinced, but definitely not worth changing back.

**Comment by tacaswell** [comment]
Do we have a theory as to why coverage looks wrong?  Are we sure that the test is actually running on CI?

**Comment by raphaelquast** [comment]
@tacaswell thanks for the follow-up!

Good point... I assumed that pytest grabs it just like any other test... however looking at codecov it seems that something strange is going on...  

The test itself is fully covered so I guess it was executed, but I can't really wrap my head around how it is even possible that the resulting coverage looks like this (in `backend_bases.py` [here](https://app.codecov.io/gh/matplotlib/matplotlib/pull/27148/blob/lib/matplotlib/backend_bases.py#L2975))...



I'll try to look into this as soon as I find some time!

**Comment by tacaswell** [comment]
At least locally for me they are being discovered, but I can reproduce the lack of coverage.  It looks like it only ever runs until the `if len(axes): return []` fast-exit.

**Comment by tacaswell** [comment]
Any update on this?

**Comment by tacaswell** [comment]
No worries!  For volunteers basically everything else in your life should take priority and we understand stuff comes up.

**Comment by ksunden** [comment]
@raphaelquast gentle ping as we are rounding up things for 3.9

**Comment by raphaelquast** [comment]
Hey @ksunden, thanks for pinging. 
It's really a pity but I'm all fed up with other things at the moment so I can't promise a timely update here.

As I see it the only thing that holds this back is the mysterious unittests that somehow don't run as expected. 
Everyting else is done (api-change-notice, stub-files, etc.). 

--- 

I quickly did another check if I can reproduce this but unfortunately I really don't get why the tests are not showing the expected coverage. 

To clarify, I just did the following crude test to be absolutely sure the lines are covered locally on my side:

- I add `print("Found axes to trigger:", axes_to_trigger)` at the bottom of the `_start_event_axes_interaction(..)` function  of `backend_bases.py` (e.g. at a position that is supposedly not covered by the tests)
- I run `python -m pytest tests -k test_interactive_pan_zoom_events -s`   to run my test (`-s` flag to see print outputs)

<details> <summary>... and this is what pytest gives me: [click to show]</summary>

```
(mpl-dev) C:\Users\Admin\Projects\matplotlib\lib\matplotlib>python -m pytest tests -k test_interactive_pan_zoom_events -s
============================================================= test session starts ==============================================================
platform win32 -- Python 3.11.6, pytest-7.4.2, pluggy-1.3.0
rootdir: C:\Users\Admin\Projects\matplotlib
configfile: pytest.ini
plugins: anyio-4.0.0, cov-4.1.0, rerunfailures-12.0, timeout-2.2.0, xdist-3.3.1
collected 9344 items / 9296 deselected / 2 skipped / 48 selected

tests\test_backend_bases.py Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t_twin'>, <Axes: label='ax_t'>, <Axes: label='ax_b_twin'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.Found axes to trigger: [<Axes: label='ax_t'>, <Axes: label='ax_b'>]
.

=============================================== 48 passed, 2 skipped, 9296 deselected in 26.49s ================================================
```

</details>

... which is what I was expecting to get if the function was executed properly

> So long story short, I unfortunately cannot reproduce the lack of coverage so I am quite stuck.  
> Since my time is very limited at the moment I don't know how to proceed here as I don't have the time to dig into reasons why > the lines are apparently executed but not reported to codecov.
> 
> Any help here is highly appreciated!  (I can try my best to finalize this for 3.9 but I'd need a way to reproduce it) 
> My last guess is that it might somehow be OS-related?? (I'm on windows)

**Comment by QuLogic** [comment]
> At least locally for me they are being discovered, but I can reproduce the lack of coverage. It looks like it only ever runs until the `if len(axes): return []` fast-exit.

I can't reproduce locally after a rebase to latest `main` (though I didn't try before the rebase); maybe that will help here?

**Comment by raphaelquast** [comment]
@QuLogic Thanks a lot for taking a look! I updated the branch to the latest `main`... let's see if that helps!

**Comment by raphaelquast** [comment]
@QuLogic I rebased to the latest `main` but now the `macOS_py311` test on azure is failing with the (I guess unrelated) error:
`Path does not exist: /Users/runner/work/1/s/result_images`

**Comment by QuLogic** [comment]
It appears that someone re-ran that, as it seems to be passing now. The code coverage looks pretty good now; only the early returns are uncovered.

**Comment by raphaelquast** [comment]
OK perfect! But `codecov/project/tests` is still failing (even though all added/changed lines are covered)?

**Review by tacaswell — APPROVED** [review]

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+To disable pan/zoom events completely, use ``ax.set_navigate(False)``
+
+"""
+
+
+import matplotlib.pyplot as plt
+
+f = plt.figure(figsize=(11, 6))
+f.suptitle("Showcase for pan/zoom events on overlapping axes.")
```
Comment (QuLogic) [inline galleries/examples/showcase/pan_zoom_overlap.py:36]: We usually use `fig`:
```suggestion
fig = plt.figure(figsize=(11, 6))
fig.suptitle("Showcase for pan/zoom events on overlapping axes.")
```

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+ax = f.add_axes((.05, .05, .9, .9))
```
Comment (QuLogic) [inline galleries/examples/showcase/pan_zoom_overlap.py:38]:
```suggestion
ax = fig.add_axes((.05, .05, .9, .9))
```

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ax.patch.set_color(".75")
+ax_twin = ax.twinx()
+
+ax1 = f.add_subplot(221)
```
Comment (QuLogic) [inline galleries/examples/showcase/pan_zoom_overlap.py:42]:
```suggestion
ax1 = fig.add_subplot(221)
```

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+ax1_twin = ax1.twinx()
+ax1.text(.5, .5,
+         "Visible patch\n\n"
+         "Pan/zoom events are NOT\n"
+         "forwarded to axes below",
+         ha="center", va="center", transform=ax1.transAxes)
+
+ax11 = f.add_subplot(223, sharex=ax1, sharey=ax1)
```
Comment (QuLogic) [inline galleries/examples/showcase/pan_zoom_overlap.py:50]:
```suggestion
ax11 = fig.add_subplot(223, sharex=ax1, sharey=ax1)
```

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ax11.set_forward_navigation_events(True)
+ax11.text(.5, .5,
+          "Visible patch\n\n"
+          "Override capture behavior:\n\n"
+          "ax.set_forward_navigation_events(True)",
+          ha="center", va="center", transform=ax11.transAxes)
+
+ax2 = f.add_subplot(222)
```
Comment (QuLogic) [inline galleries/examples/showcase/pan_zoom_overlap.py:58]:
```suggestion
ax2 = fig.add_subplot(222)
```

File: galleries/examples/showcase/pan_zoom_overlap.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+ax2_twin = ax2.twinx()
+ax2.patch.set_visible(False)
+ax2.text(.5, .5,
+         "Invisible patch\n\n"
+         "Pan/zoom events are\n"
+         "forwarded to axes below",
+         ha="center", va="center", transform=ax2.transAxes)
+
+ax22 = f.add_subplot(224, sharex=ax2, sharey=ax2)
```
Comment (QuLogic) [inline galleries/examples/showcase/pan_zoom_overlap.py:67]:
```suggestion
ax22 = fig.add_subplot(224, sharex=ax2, sharey=ax2)
```

File: lib/matplotlib/backend_bases.py [hunk 12]
```diff
@@ -2995,6 +2995,59 @@ def _zoom_pan_handler(self, event):
             elif event.name == "button_release_event":
                 self.release_zoom(event)
 
+    def _start_event_axes_interaction(self, event, method="zoom"):
```
Comment (QuLogic) [inline lib/matplotlib/backend_bases.py:2998]: Never called without an argument, so make it keyword-only.
```suggestion
    def _start_event_axes_interaction(self, event, *, method):
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 13]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+    init_xlim, init_ylim = (0, 10), (0, 10)
+    for ax in [ax_t, ax_b]:
+        ax.set_xlim(*init_xlim)
+        ax.set_ylim(*init_ylim)
+
+    # Mouse from 2 to 1 (in data-coordinates of ax_t).
+    xstart_t, xstop_t, ystart_t, ystop_t = 1, 2, 1, 2
+    # Convert to screen coordinates ("s").  Events are defined only with pixel
+    # precision, so round the pixel values, and below, check against the
+    # corresponding xdata/ydata, which are close but not equal to d0/d1.
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:495]: Typo?
```suggestion
    # corresponding xdata/ydata, which are close but not equal to s0/s1.
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 14]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+
+    # Calculate the mouse-distance in data-coordinates of the bottom-axes
+    xstart_b, ystart_b = ax_b.transData.inverted().transform(s0)
+    xstop_b, ystop_b = ax_b.transData.inverted().transform(s1)
+
+    # Set up the mouse movements
+    start_event = MouseEvent(
+        "button_press_event", fig.canvas, *s0, button)
+    stop_event = MouseEvent(
+        "button_release_event", fig.canvas, *s1, button)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:507]: Fits in a line:
```suggestion
    start_event = MouseEvent("button_press_event", fig.canvas, *s0, button)
    stop_event = MouseEvent("button_release_event", fig.canvas, *s1, button)
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 15]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+
+    tb = NavigationToolbar2(fig.canvas)
+
+    if tool == "zoom":
+        # Evaluate expected limits before executing the zoom-event
+        direction = ("in" if button == 1 else "out")
+
+        xlim_t, ylim_t = ax_t._prepare_view_from_bbox(
+            [*s0, *s1], direction)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:516]:
```suggestion
        xlim_t, ylim_t = ax_t._prepare_view_from_bbox([*s0, *s1], direction)
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 16]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+            xlim_b, ylim_b = ax_b._prepare_view_from_bbox(
+                [*s0, *s1], direction)
+        elif ax_t.get_forward_navigation_events() is False:
+            xlim_b = init_xlim
+            ylim_b = init_ylim
+        else:
+            if ax_t.patch.get_visible() is False:
+                xlim_b, ylim_b = ax_b._prepare_view_from_bbox(
+                    [*s0, *s1], direction)
+            else:
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:528]:
```suggestion
            xlim_b, ylim_b = ax_b._prepare_view_from_bbox([*s0, *s1], direction)
        elif ax_t.get_forward_navigation_events() is False:
            xlim_b = init_xlim
            ylim_b = init_ylim
        else:
            if not ax_t.patch.get_visible():
                xlim_b, ylim_b = ax_b._prepare_view_from_bbox([*s0, *s1], direction)
            else:
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 17]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+        assert ax_t.get_xlim() == pytest.approx(xlim_t, abs=0.15)
+        assert ax_t.get_ylim() == pytest.approx(ylim_t, abs=0.15)
+        assert ax_b.get_xlim() == pytest.approx(xlim_b, abs=0.15)
+        assert ax_b.get_ylim() == pytest.approx(ylim_b, abs=0.15)
+
+        # Check if twin-axes are properly triggered
+        assert ax_t.get_xlim() == pytest.approx(ax_t_twin.get_xlim(),
+                                                abs=0.15)
+        assert ax_b.get_xlim() == pytest.approx(ax_b_twin.get_xlim(),
+                                                abs=0.15)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:546]:
```suggestion
        assert ax_t.get_xlim() == pytest.approx(ax_t_twin.get_xlim(), abs=0.15)
        assert ax_b.get_xlim() == pytest.approx(ax_b_twin.get_xlim(), abs=0.15)
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 18]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+    else:
+        # Evaluate expected limits
+        # (call start_pan to make sure ax._pan_start is set)
+        ax_t.start_pan(*s0, button)
+        xlim_t, ylim_t = ax_t._get_pan_points(
+            button, None, *s1).T.astype(float)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:552]:
```suggestion
        xlim_t, ylim_t = ax_t._get_pan_points(button, None, *s1).T.astype(float)
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 19]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+        ax_t.end_pan()
+
+        if ax_t.get_forward_navigation_events() is True:
+            ax_b.start_pan(*s0, button)
+            xlim_b, ylim_b = ax_b._get_pan_points(
+                button, None, *s1).T.astype(float)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:558]:
```suggestion
            xlim_b, ylim_b = ax_b._get_pan_points(button, None, *s1).T.astype(float)
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 20]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+            ax_b.end_pan()
+        elif ax_t.get_forward_navigation_events() is False:
+            xlim_b = init_xlim
+            ylim_b = init_ylim
+        else:
+            if ax_t.patch.get_visible() is False:
+                ax_b.start_pan(*s0, button)
+                xlim_b, ylim_b = ax_b._get_pan_points(
+                    button, None, *s1).T.astype(float)
+                ax_b.end_pan()
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:568]:
```suggestion
            if not ax_t.patch.get_visible():
                ax_b.start_pan(*s0, button)
                xlim_b, ylim_b = ax_b._get_pan_points(button, None, *s1).T.astype(float)
                ax_b.end_pan()
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 21]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
+        assert ax_t.get_xlim() == pytest.approx(xlim_t, abs=0.15)
+        assert ax_t.get_ylim() == pytest.approx(ylim_t, abs=0.15)
+        assert ax_b.get_xlim() == pytest.approx(xlim_b, abs=0.15)
+        assert ax_b.get_ylim() == pytest.approx(ylim_b, abs=0.15)
+
+        # Check if twin-axes are properly triggered
+        assert ax_t.get_xlim() == pytest.approx(ax_t_twin.get_xlim(),
+                                                abs=0.15)
+        assert ax_b.get_xlim() == pytest.approx(ax_b_twin.get_xlim(),
+                                                abs=0.15)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:587]:
```suggestion
        assert ax_t.get_xlim() == pytest.approx(ax_t_twin.get_xlim(), abs=0.15)
        assert ax_b.get_xlim() == pytest.approx(ax_b_twin.get_xlim(), abs=0.15)
```

File: lib/matplotlib/tests/test_backend_bases.py [hunk 22]
```diff
@@ -439,3 +439,149 @@ def test_toolmanager_update_keymap():
...
         fig.canvas.manager.toolmanager.update_keymap('foo', 'c')
+
+
+@pytest.mark.parametrize("tool", ["zoom", "pan"])
+@pytest.mark.parametrize("button", [MouseButton.LEFT, MouseButton.RIGHT])
+@pytest.mark.parametrize("patch_vis", [True, False])
+@pytest.mark.parametrize("forward_nav", [True, False, "auto"])
+@pytest.mark.parametrize("t_s", ["twin", "share"])
+def test_interactive_pan_zoom_events(tool, button, patch_vis,
+                                     forward_nav, t_s):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_bases.py:450]:
```suggestion
def test_interactive_pan_zoom_events(tool, button, patch_vis, forward_nav, t_s):
```

**Review by QuLogic — APPROVED** [review]

**Comment by ksunden** [comment]
GTK dll loading failures on appveyor appears to be unrelated, reproducible in other recent PRs.

**Comment by ksunden** [comment]
Thanks for your perseverance on this one, it was a long journey, but glad we got there!

**Comment by raphaelquast** [comment]
Awesome to see that this is finally merged! 
I learned a lot on the road and I'm really happy that this journey has come to a successful ending.
Thanks also from my side for all the help along the way!

**Comment by tacaswell** [comment]
I hope we hear from you again @raphaelquast !

**Comment by raphaelquast** [comment]
@tacaswell definitely! I do have a few things that I'd like to see implemented and I hope that I find some time to contribute!
